src/repl/scala/tools/nsc/interpreter/IMain.scala

```diff
@@ -316,14 +316,14 @@ class IMain(initialSettings: Settings, protected val out: JPrintWriter) extends
     def shift[T](op: => T): T = exitingFlatten(op)
   }
 
-  def originalPath(name: String): String = originalPath(TermName(name))
-  def originalPath(name: Name): String   = translateOriginalPath(typerOp path name)
-  def originalPath(sym: Symbol): String  = translateOriginalPath(typerOp path sym)
+  def originalPath(name: String, definesTopLevel: Boolean): String = originalPath(TermName(name), definesTopLevel)
+  def originalPath(name: Name, definesTopLevel: Boolean): String   = translateOriginalPath(typerOp path name, definesTopLevel)
+  def originalPath(sym: Symbol, definesTopLevel: Boolean): String  = translateOriginalPath(typerOp path sym, definesTopLevel)
 
   /** For class based repl mode we use an .INSTANCE accessor. */
-  def readInstanceName = if (isClassBased) ".INSTANCE" else ""
-  def translateOriginalPath(p: String): String = {
-    if (isClassBased) p.replace(sessionNames.read, sessionNames.read + readInstanceName) else p
+  def readInstanceName(definesTopLevel: Boolean) = if (isClassBased && !definesTopLevel) ".INSTANCE" else ""
```

Comment:
As discussed, we might be able to add `def INSTANCE = this` to the object based wrappers and then just always emit the same path here.